src/unauthenticated_stream/chacha20.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+use {chacha, error, polyfill, unauthenticated_stream};
+use polyfill::slice::u32_from_le_u8;
+
+/// ChaCha20 unauthenticated stream cipher.
+///
+/// The keys are 256 bits long and the nonces are 128 bits long, rather then
+/// the usual 96 bits. The first 32 bits of the 128-bit nonce represents an
+/// explicit initial counter parameter.
+pub static CHACHA20: unauthenticated_stream::Algorithm =
+                                        unauthenticated_stream::Algorithm {
+    key_len: chacha::KEY_LEN_IN_BYTES,
+    init: chacha20_init,
+    xor_keystream: chacha20_xor_keystream,
+    id: unauthenticated_stream::AlgorithmID::CHACHA20,
+    max_input_len: max_input_len!(CHACHA20_BLOCK_LEN, CHACHA20_OVERHEAD_BLOCKS_PER_NONCE),
+};
+
+const CHACHA20_BLOCK_LEN: u64 = 64;
+const CHACHA20_OVERHEAD_BLOCKS_PER_NONCE: u64 = 1;
+
+/// Copies |key| into |ctx_buf|.
+pub fn chacha20_init(ctx_buf: &mut [u8], key: &[u8])
+                              -> Result<(), error::Unspecified> {
+    ctx_buf[..key.len()].copy_from_slice(key);
+    Ok(())
+}
+
+fn chacha20_xor_keystream(ctx: &[u64; unauthenticated_stream::KEY_CTX_BUF_ELEMS],
+                          nonce: &[u8; unauthenticated_stream::NONCE_LEN],
+                          in_out: &mut [u8]) -> Result<(), error::Unspecified> {
+    let chacha20_key = ctx_as_key(ctx)?;
+    let mut counter = make_counter(nonce);
+    chacha::chacha20_xor_in_place(&chacha20_key, &counter, in_out);
+    counter[0] = 0;
+    Ok(())
```

Comment:
This is exactly `chacha20_poly1305_seal()` except `chacha20_poly1305_seal()` also does a Poly1305 operation. `chacha20_poly1305_seal()` should be re-implemented in terms of `chacha20_xor_keystream`.

I still don't like the `_keystream` in the name. Let's just remove or replace that. Let's add `_in_place` to the name so that the reader of the calling code can more quickly understand that the input buffer is also the output buffer.

Reply:
ACK.

Reply:
Actually, there is another difference between this and `chacha20_poly1305_seal`: since the aead module uses 12 bytes nonces with an hardocded initial counter, while this uses the counter as passed by the application as part of a 16 bytes nonce (this specifically is needed for QUIC).

Though `chacha20_poly1305_seal` can still be implemented by using `chacha20_xor_keystream` by just making sure it passed the 16 bytes nonce as returned by `make_counter`.

Reply:
Right, though the AES-CTR case uses a 16 byte nonce and doesn't mention an explicit counter, so in order to make a generic API we'd either have to make the application pass the explicit counter to encrypt_in_place/decrypt_in_place (and let the application figure out what that means for AES-CTR), or pass a 16 bytes nonce in both cases (and let the application figure out what 16 bytes nonce means for ChaCha20, although the solution is in the text you linked to) which is what I did here.

The latter is somewhat easier from the application POV since ring already has support for turning a 4 byte slice into a u32, while the application would need to implement it on top of using ring's API.

But I can look into doing the explicit counter option and see how much more difficult it is to use vs. how much better it makes the internal ring code.

Reply:
> Right, though the AES-CTR case uses a 16 byte nonce and doesn't mention an explicit counter,

I see. Thanks for explaining that.

For now, I suggest we expose the AES-CTR code as taking a 96-bit nonce and a 32-bit counter, like how QUIC specifies for ChaCah20. Internally *ring*'s AES-CTR implementation only supports incrementing the last 32 bits of the counter so any other style of API (e.g. one that takes a 128-bit initial counter) would be misleading as far as the capabilities go.

Reply:
So, trying to figure out how that would work in an application, the problem I see is that that would make having a common API for both ChaCha20 and AES-CTR pointless, because the application would need to be aware of the specific algorithm being used to know how to generate different counters based on the different algorithms (i.e. for ChaCha it would use the first 4 bytes, while for AES-CTR it's actually the last 4 AFAICT).

So, if we go with the explicit counter, I'd say we might as well just expose the raw "ChaCha20()" and "AES-CTR()" operations directly as used in the QUIC spec without having to go to all the trouble of making a generic API which wouldn't really make the application's life easier.

The application would need to generate different key objects depending on the algorithm (so would need to expose key generation APIs as well for both ChaCha and AES), which I think is fine since this isn't intended for general use anyway, and it would discourage use unless really needed.

Reply:
> So, trying to figure out how that would work in an application, the problem I see is that that would make having a common API for both ChaCha20 and AES-CTR pointless,

I agree that it would be less than ideal, but I "pointless" is an overstatement.

> because the application would need to be aware of the specific algorithm being used to know how to generate different counters based on the different algorithms (i.e. for ChaCha it would use the first 4 bytes, while for AES-CTR it's actually the last 4 AFAICT).

With your original proposed API, the application would need to know these details too. It happens that for a QUIC implementation, the details just happen to work for what QUIC needs, but there would be no way for somebody other than us to know that given the API signature and the API documentation given in this PR to this point.

For QUIC, the details of the counter format matter less because QUIC only needs to encrypt one block (IIUC) so the counter never gets incremented during the operation. However, for non-QUIC applications that would use this API, the details of how the counter is constructed (big-endian for AES-GCM, little-endian for ChaCha20, IIUC) and how it is incremented (mod 2**32) and the amount of data that can be encrypted in a single operation (2**32 * block size, approximately) matter a lot.

Now, that said, I also eventually want to extend this API to work for non-CTR modes, in particular CBC mode. For those modes a 96-bit nonce and a 32-bit counter as separate arguments don't make sense. So I think it would be OK to keep the API in the same shape you presented it, especially if the difference in the counter/nonce split between ChaCha20 and AES-GCM is explained by the little-endian vs. big-endian nature of each; Do you agree that the difference comes down to endianness, or is it something different?

Reply:
> So, trying to figure out how that would work in an application, the problem I see is that that would make having a common API for both ChaCha20 and AES-CTR pointless,

I agree that it would be less than ideal, but I "pointless" is an overstatement.

> because the application would need to be aware of the specific algorithm being used to know how to generate different counters based on the different algorithms (i.e. for ChaCha it would use the first 4 bytes, while for AES-CTR it's actually the last 4 AFAICT).

With your original proposed API, the application would need to know these details too. It happens that for a QUIC implementation, the details just happen to work for what QUIC needs, but there would be no way for somebody other than us to know that given the API signature and the API documentation given in this PR to this point.

For QUIC, the details of the counter format matter less because QUIC only needs to encrypt one block (IIUC) so the counter never gets incremented during the operation. However, for non-QUIC applications that would use this API, the details of how the counter is constructed (big-endian for AES-GCM, little-endian for ChaCha20, IIUC) and how it is incremented (mod 2**32) and the amount of data that can be encrypted in a single operation (2**32 * block size, approximately) matter a lot.

Now, that said, I also eventually want to extend this API to work for non-CTR modes, in particular CBC mode. For those modes a 96-bit nonce and a 32-bit counter as separate arguments don't make sense. So I think it would be OK to keep the API in the same shape you presented it, especially if the difference in the counter/nonce split between ChaCha20 and AES-GCM is explained by the little-endian vs. big-endian nature of each; Do you agree that the difference comes down to endianness, or is it something different?

Reply:
> So, trying to figure out how that would work in an application, the problem I see is that that would make having a common API for both ChaCha20 and AES-CTR pointless,

I agree that it would be less than ideal, but I "pointless" is an overstatement.

> because the application would need to be aware of the specific algorithm being used to know how to generate different counters based on the different algorithms (i.e. for ChaCha it would use the first 4 bytes, while for AES-CTR it's actually the last 4 AFAICT).

With your original proposed API, the application would need to know these details too. It happens that for a QUIC implementation, the details just happen to work for what QUIC needs, but there would be no way for somebody other than us to know that given the API signature and the API documentation given in this PR to this point.

For QUIC, the details of the counter format matter less because QUIC only needs to encrypt one block (IIUC) so the counter never gets incremented during the operation. However, for non-QUIC applications that would use this API, the details of how the counter is constructed (big-endian for AES-GCM, little-endian for ChaCha20, IIUC) and how it is incremented (mod 2**32) and the amount of data that can be encrypted in a single operation (2**32 * block size, approximately) matter a lot.

Now, that said, I also eventually want to extend this API to work for non-CTR modes, in particular CBC mode. For those modes a 96-bit nonce and a 32-bit counter as separate arguments don't make sense. So I think it would be OK to keep the API in the same shape you presented it, especially if the difference in the counter/nonce split between ChaCha20 and AES-GCM is explained by the little-endian vs. big-endian nature of each; Do you agree that the difference comes down to endianness, or is it something different?

> The application would need to generate different key objects depending on the algorithm (so would need to expose key generation APIs as well for both ChaCha and AES), which I think is fine since this isn't intended for general use anyway, and it would discourage use unless really needed.

I don't think we need key generation APIs for QUIC since QUIC uses derived keys, right?

Reply:
> So, trying to figure out how that would work in an application, the problem I see is that that would make having a common API for both ChaCha20 and AES-CTR pointless,

I agree that it would be less than ideal, but I "pointless" is an overstatement.

> because the application would need to be aware of the specific algorithm being used to know how to generate different counters based on the different algorithms (i.e. for ChaCha it would use the first 4 bytes, while for AES-CTR it's actually the last 4 AFAICT).

With your original proposed API, the application would need to know these details too. It happens that for a QUIC implementation, the details just happen to work for what QUIC needs, but there would be no way for somebody other than us to know that given the API signature and the API documentation given in this PR to this point.

For QUIC, the details of the counter format matter less because QUIC only needs to encrypt one block (IIUC) so the counter never gets incremented during the operation. However, for non-QUIC applications that would use this API, the details of how the counter is constructed (big-endian for AES-GCM, little-endian for ChaCha20, IIUC) and how it is incremented (mod 2**32) and the amount of data that can be encrypted in a single operation (2**32 * block size, approximately) matter a lot.

Now, that said, I also eventually want to extend this API to work for non-CTR modes, in particular CBC mode. For those modes a 96-bit nonce and a 32-bit counter as separate arguments don't make sense. So I think it would be OK to keep the API in the same shape you presented it, especially if the difference in the counter/nonce split between ChaCha20 and AES-GCM is explained by the little-endian vs. big-endian nature of each; Do you agree that the difference comes down to endianness, or is it something different?

> The application would need to generate different key objects depending on the algorithm (so would need to expose key generation APIs as well for both ChaCha and AES), which I think is fine since this isn't intended for general use anyway, and it would discourage use unless really needed.

I don't think we need key generation APIs for QUIC since QUIC uses derived keys, right?

Reply:
> Do you agree that the difference comes down to endianness, or is it something different?

Yesh, makes sense.

> I don't think we need key generation APIs for QUIC since QUIC uses derived keys, right?

The AES key still needs the whole initialization step AFAICT (that's what I meant with "key generation").